---
layout: default
title: 스프링 트랜잭션과 예외-1
grand_parent: 스프링
parent: 스프링 트랜잭션
nav_order: 1
---

# 검색영역
{: .no_toc }

## 목차는 여기인 것 같아요.
{: .no_toc .text-delta }

1. TOC
{:toc}

---

## 스프링 트랜잭션 예외별 커밋, 롤백 확인

예외가 발생했는데, 내부에서 예외를 처리하지 못하고, 트랜잭션 범위(aop) 밖으로 예외를 던지면 어떻게 될까?
예외 발생시 스프링 트랜잭션 AOP는 예외의 종류에 따라 트랜잭션을 커밋하거나 롤백한다.

- 언체크 예외인 체크 예외인 RuntimeException, Error와 그 하위 예외가 발생하면 트랜잭션을 롤백한다.
- 체크 예외인 Exception 과 그 하위 예외가 발생하면 트랜잭션을 커밋한다.
- 정상 응답(리턴)하면 트랜잭션을 커밋한다.

```java
@SpringBootTest
public class RollbackTest {

    @Autowired RollbackService service;

    @Test
    void runtimeException() {
        assertThatThrownBy(() -> service.runtimeException()).isInstanceOf(RuntimeException.class);
    }

    @Test
    void checkedException() {
        assertThatThrownBy(() -> service.checkedException()).isInstanceOf(MyException.class);
    }

    @Test
    void rollbackFor() {
        assertThatThrownBy(() -> service.rollbackFor()).isInstanceOf(MyException.class);
    }

    @TestConfiguration
    static class RollbackTestConfig {
        @Bean
        RollbackService rollbackService() {
            return new RollbackService();
        }
    }

    @Slf4j
    static class RollbackService {
        //런타임 예외 발생: 롤백
        @Transactional
        public void runtimeException() {
            log.info("call runtimeException");
            throw new RuntimeException();
        }

        //체크 예외 발생: 커밋
        @Transactional
        public void checkedException() throws MyException {
            log.info("call checkedException");
            throw new MyException();
        }

        //체크 예외 rollbackFor 지정: 롤백
        @Transactional(rollbackFor = MyException.class)
        public void rollbackFor() throws MyException {
            log.info("call rollbackFor");
            throw new MyException();
        }
    }

    static class MyException extends Exception {

    }
}
```

```
logging.level.org.springframework.transaction.interceptor=TRACE
logging.level.org.springframework.jdbc.datasource.DataSourceTransactionManager=DEBUG
#JPA log
logging.level.org.springframework.orm.jpa.JpaTransactionManager=DEBUG
logging.level.org.hibernate.resource.transaction=DEBUG
```

application.properties 에 해당 옵션을 주면 트랜잭션이 커밋, 롤백되는지 확인할 수 있다.

## 예외와 트랜잭션 커밋, 롤백 - 활용

스프링은 왜 체크 예외는 커밋하고, 언체크(런타임) 예외는 롤백할까?
스프링 기본적으로 체크 예외는 비즈니스 의미가 있을 때 사용하고, 런타임(언체크) 예외는 복구 불가능한 예외로 가정한다.

- 체크 예외: 비즈니스 의미가 있을 때 사용
- 언체크 예외: 복구 불가능한 예외

꼭 이런 정책을 따를 필요는 없다. rollbackFor 옵션 사용해서 체크 예외도 롤백하면 된다.

그런데 비즈니스 의미가 있는 비즈니스 예외와 복구 불가능한 예외라는 것이 무슨 뜻일까?

```
복구 불가능한 예외
sql문법오류, 통신이 안됨같은 경우는 복구가 불가능하다. 이런 예외는 catch 하더라도 딱히 방법이 없다. 공통부에서 처리를 해야한다.
```

```
비즈니스 요구사항
주문을 하는데 상황에 따라 다음과 같이 조치한다.

- 1. 정상 : 주문시 결제를 성공하면 주문 데이터를 저장하고 결제 상태를 완료로 처리한다.
- 2. 시스템 예외 : 주문시 내부에 복구 불가능한 예외가 발생하면 전체 데이터를 롤백한다.
- 3. 비즈니스 예외 : 주문시 결제 잔고가 부족하면 주문 데이터를 저장하고, 결제 상태를 대기로 처리한다. 이 경우 고객에게 잔고 부족을 알리고 별도의 계좌로 입금하도록 안내한다.
```

이때 결제 잔고가 부족하면 NotEnoughMoneyException이라는 체크 예외가 발생한다고 가정하겠다. 
이 예외는 시스템에 문제가 있어서 발생하는 시스템 예외가 아니다. 시스템은 정상 동작했지만, 
비즈니스 상황에서 문제가 되기 때문에 발생한 예외이다. 
더 자세히 설명하자면, 고객의 잔고가 부족한 것은 시스템에 문제가 있는 것이 아니다. 
오히려 시스템은 문제 없이 동작한 것이고, 비즈니스 상황이 예외인 것이다. 이런 예외를 비즈니스 예외라 한다. 
그리고 비즈니스 예외는 매우 중요하고, 반드시 처리해야 하는 경우가 많으므로 체크 예외를 고려할 수 있다.















